lib/disk/modules/VMWareSparseDisk.rb

```diff
@@ -2,49 +2,117 @@
 require 'disk/modules/miq_disk_common'
 require 'binary_struct'
 require 'memory_buffer'
+require 'zlib'
 
 module VMWareSparseDisk
+  SPARSE_MAGIC_NUMBER = 0x564d444b # 'V' 'M' 'D' 'K'
+
   SPARSE_EXTENT_HEADER = BinaryStruct.new([
-    'L', 'magicNumber',
-    'L', 'version',
-    'L', 'flags',
-    'Q', 'capacity',
-    'Q', 'grainSize',
-    'Q', 'descriptorOffset',
-    'Q', 'descriptorSize',
-    'L', 'numGTEsPerGT',
-    'Q', 'rgdOffset',
-    'Q', 'gdOffset',
-    'Q', 'overHead',
-    'C', 'uncleanShutdown'
+    # the magic number is used to verify the validity of each sparse extent when the extent is opened
+    'L',  'magicNumber',          # should be SPARSE_MAGIC_NUMBER
+
+    'L',  'version',              # the version number can be 1 or 2
+
+    # SparseExtentHeader is stored on disk in little-endian byte order, so if you examine the first eight bytes of a VMDK file,
+    # you see either:
+    #   'K' D' 'M' 'V' 0x01 0x00 0x00 0x00
+    #   'K' D' 'M' 'V' 0x02 0x00 0x00 0x00
+
+    # flags contains the following bits of information.
+    'L',  'flags',                # bit 0:  valid new line detection test
```

Comment:
I think there's a way to do bit level here, so you can actually give each of these flags an individual name here, and then you don't need the method that maps to a hash.

Reply:
I tried doing this, but all the ways to do it are kind of clunky, so probably ok to go with what's here...

So I thought you could do 

```ruby
'b', :flag1,
'b', :flag2,
'b', :flag3,
...
```

Unfortunately the way the `b` identifier works is that it consume 1 but then moves the iterator forward the rest of the byte.  So, the only way to read consecutive bits is to do something like...

```ruby
'b8', :flags
```

which then gives something like `{:flags => "01100100"}`.

---

I think BinaryStruct could be enhanced with some sort of bit decoding, but I'm not sure what a nice interface would look like offhand.  I'm guessing the caller would just want to deal with bits individually, but under the covers as we build the format string we'd have to understand consecutive `b` definitions and pack them into a single `b<N>`.

Reply:
That being said, it *might* be easier to use the bit decoding and do something like this...

```ruby
# Instead of
'L',  'flags'

# use this
'b16' 'flags'

# and then immediately after decoding do:

sparseHeader["flags"] = sparseHeader["flags"].chars.map {|c| c == "1"}

# Then perhps something like...

sparseHeader["valid_new_line_detection_test"] = sparseHeader["flags"][0]
sparseHeader["redundant_grain_table_used"]    = sparseHeader["flags"][1]
sparseHeader["zero_grain_gte_used"]           = sparseHeader["flags"][2]
sparseHeader["grains_compressed"]             = sparseHeader["flags"][16]

Reply:
shouldn't this be `b32` since we need to access bit 0 as well as bit 16 and bit 17?

Reply:
Oh yes...I typoed it.